gateway/coprocess_bundle_test.go

```diff
@@ -243,3 +247,55 @@ func TestResponseOverride(t *testing.T) {
 		testOverride(t, ts.RegisterBundle("jsvm_override", overrideResponseJSVM))
 	})
 }
+
+func TestPullBundle(t *testing.T) {
+
+	testCases := []struct {
+		name             string
+		expectedAttempts int
+		shouldErr        bool
+	}{
+		{
+			name:             "bundle downloaded at first attempt",
+			expectedAttempts: 1,
+			shouldErr:        false,
+		},
+		{
+			// failed the 2 first times
+			name:             "bundle downloaded at third attempt",
+			expectedAttempts: 3,
+			shouldErr:        false,
+		},
+		{
+			// should try 5 times, afterwards it will fail
+			name:             "bundle download failed",
+			expectedAttempts: 5,
+			shouldErr:        true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			attempts := 0
+
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				attempts++
+				if tc.expectedAttempts > attempts || tc.shouldErr {
+					// simulate file not found, so it will throw err
+					w.WriteHeader(http.StatusNotFound)
+				}
+			}))
+			defer ts.Close()
+
+			getter := &HTTPBundleGetter{
+				URL:                ts.URL,
+				InsecureSkipVerify: false,
+			}
+			_, err := pullBundle(getter)
```

Comment:
pullBundle uses the backoff defined internally, must be overriden to 0, to avoid sleeping from tests. We don't want tests that take seconds to complete.

Reply:
--- PASS: TestPullBundle/bundle_downloaded_at_first_attempt (0.00s)
--- PASS: TestPullBundle/bundle_downloaded_at_third_attempt (0.52s)
--- PASS: TestPullBundle/bundle_download_failed (1.10s)

The default retries are also too short. The package default starts with 500ms.

Reply:
@titpetric go it! will modify the multiplier. Regarding tests, I am not considering the time spent between one attempt and another, and just considering the number of hits that I expect to receive (Eg: will fail the first 2 times, then I must receive a 3rd hit)

Reply:
good catch!! updated